jwt/jwt.go

```diff
@@ -0,0 +1,259 @@
+package jwt
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+const (
+	defaultLeewaySeconds = 150
+)
+
+// Validator validates JSON Web Tokens (JWT) by providing signature
+// verification and claims set validation.
+type Validator struct {
+	keySet KeySet
+}
+
+// NewValidator returns a Validator that uses the given KeySet to verify JWT signatures.
+func NewValidator(keySet KeySet) (*Validator, error) {
+	if keySet == nil {
+		return nil, errors.New("keySet must not be nil")
+	}
+
+	return &Validator{
+		keySet: keySet,
+	}, nil
+}
+
+// Expected defines the expected claims values to assert when validating a JWT.
+// For claims that involve validation of the JWT with respect to time, leeway
+// fields are provided to account for potential clock skew.
+type Expected struct {
+	// The expected JWT "iss" (issuer) claim value. If empty, validation is skipped.
+	Issuer string
+
+	// The expected JWT "sub" (subject) claim value. If empty, validation is skipped.
+	Subject string
+
+	// The expected JWT "jti" (JWT ID) claim value. If empty, validation is skipped.
+	ID string
+
+	// The list of expected JWT "aud" (audience) claim values to match against.
+	// The JWT claim will be considered valid if it matches any of the expected
+	// audiences. If there are no expected audiences, then the presence of any
+	// audience in the JWT will render it invalid.
+	Audiences []string
+
+	// SigningAlgorithms provides the list of expected JWS "alg" (algorithm) header
```

Comment:
Discussed moving [oidc/algs.go](https://github.com/hashicorp/cap/blob/main/oidc/algs.go) to a common package and restricting `Expected.SigningAlgorithms` to the set of algorithms defined there. I'm happy to make that change if there is agreement on that.

Reply:
I've made changes to restrict the set of expected `SigningAlgorithms` in d8a696d. I decided to copy the constants from [oidc/algs.go](https://github.com/hashicorp/cap/blob/main/oidc/algs.go) into the jwt package instead of creating a "common" or "shared" package. This approach seemed okay for constants. I'm happy to consider alternative ideas around adding them to a separate package though.